### PR TITLE
Statsbeat for GraalVM native images

### PR DESCRIPTION
**Pull request by jeanbisutti**
This PR:
* Enable Statsbeat by default in the standalone exporter
* Add a feature to detect GraalVM native usage
* Make the Statsbeat features work in a GraalVM native environment
* Track additional instrumentations (automatic instrumentation of the OTel starter and OTel instrumentation libraries)

### CODE REVIEW

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorExporterBuilder.java [hunk 1]
```diff
@@ -376,10 +380,17 @@ private MetricExporter buildMetricExporter(ConfigProperties configProperties) {
     }
 
     private Set<Feature> initStatsbeatFeatures() {
-        // TODO (jean): start tracking native image usage based on a system property or env var to indicate it's from the native image path
+        if (isGraalVmNativeExecution()) {
```
Comment (heyams) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorExporterBuilder.java:383]:
```suggestion
        if (System.getProperty("org.graalvm.nativeimage.imagecode") != null) {
```

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorExporterBuilder.java [hunk 2]
```diff
@@ -376,10 +380,17 @@ private MetricExporter buildMetricExporter(ConfigProperties configProperties) {
...
+            return Collections.singleton(Feature.GRAAL_VM_NATIVE);
+        }
         return Collections.emptySet();
     }
 
+    private static boolean isGraalVmNativeExecution() {
+        String imageCode = System.getProperty("org.graalvm.nativeimage.imagecode");
+        return imageCode != null;
+    }
```
Comment (heyams) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorExporterBuilder.java:392]: inline

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/Instrumentations.java [hunk 3]
```diff
@@ -95,6 +95,17 @@ public class Instrumentations {
...
+        INSTRUMENTATION_MAP.put("io.opentelemetry.micrometer-1.5", 73);
+        INSTRUMENTATION_MAP.put("io.opentelemetry.kafka-clients-2.6", 74);
+        INSTRUMENTATION_MAP.put("io.opentelemetry.spring-kafka-2.7", 75);
+        INSTRUMENTATION_MAP.put("io.opentelemetry.spring-web-3.1", 76);
+        INSTRUMENTATION_MAP.put("io.opentelemetry.spring-webmvc-5.3", 77);
+        INSTRUMENTATION_MAP.put("io.opentelemetry.spring-webmvc-6.0", 78);
+        INSTRUMENTATION_MAP.put("io.opentelemetry.spring-webflux-5.3", 79);
+        INSTRUMENTATION_MAP.put("io.opentelemetry.oshi", 80);
+        INSTRUMENTATION_MAP.put("io.opentelemetry.runtime-telemetry-java8", 81);
+        INSTRUMENTATION_MAP.put("io.opentelemetry.runtime-telemetry-java17", 82);
```
Comment (heyams) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/Instrumentations.java:108]: how do these relate to GraalVM?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorTraceExporter.java [hunk 4]
```diff
@@ -62,6 +66,19 @@ public CompletableResultCode export(Collection<SpanData> spans) {
...
+            return;
+        }
+        String instrumentationScopeName = span.getInstrumentationScopeInfo().getName();
+        boolean isAnAzureLibraryInstrumentation = instrumentationScopeName.startsWith(
+            "azure-");
+        if (isAnAzureLibraryInstrumentation) {
+            instrumentationScopeName = AZURE_OPENTELEMETRY;
+        }
+        statsbeatModule.getInstrumentationStatsbeat().addInstrumentation(instrumentationScopeName);
+    }
```
Comment (heyams) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorTraceExporter.java:80]: this is not related to GraalVM, right? i think we need to discuss this first.

**Comment by heyams** [comment]
> This PR:
> 
**> * Enable Statsbeat by default in the standalone exporter**
> * Add a feature to detect GraalVM native usage
> * Make the Statsbeat features work in a GraalVM native environment
> * Track additional instrumentations (automatic instrumentation of the OTel starter and OTel instrumentation libraries)

what do you mean #1 above? it has statsbeat enabled by default for standalone exporter. Attach and some features, java.vendor.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/Instrumentations.java (hunk 3 above)
Comment (jeanbisutti) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/Instrumentations.java:108]: Automatic instrumentation of the OTel starter and OTel libraries can be used with GraalVM native

**Comment by jeanbisutti** [comment]
> > This PR:
> 
> **> * Enable Statsbeat by default in the standalone exporter**
> 
> > * Add a feature to detect GraalVM native usage
> > * Make the Statsbeat features work in a GraalVM native environment
> > * Track additional instrumentations (automatic instrumentation of the OTel starter and OTel instrumentation libraries)
> 
> what do you mean #1 above? it has statsbeat enabled by default for standalone exporter. Attach and some features, java.vendor.

To be more accurate, the PR enables the instrumentation tracking by default

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorTraceExporter.java (hunk 4 above)
Comment (jeanbisutti) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorTraceExporter.java:80]: With GraalVM native, there are 2 main options today: use [this Spring Boot-related project](https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/spring/spring-cloud-azure-starter-monitor) or the standalone exporter with instrumentation libraries or Quarkus.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorTraceExporter.java [hunk 5]
```diff
@@ -62,6 +66,19 @@ public CompletableResultCode export(Collection<SpanData> spans) {
         return telemetryItemExporter.send(telemetryItems);
     }
 
+    private void addInstrumentationToStatBeat(SpanData span) {
```
Comment (heyams) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorTraceExporter.java:69]:
```suggestion
    private void addInstrumentationToStatsbeat(SpanData span) {
```

Comment (heyams) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorTraceExporter.java:69]: are you going to turn this into a utility method so that it can be reused for Java agent?

Comment (jeanbisutti) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorTraceExporter.java:69]: I had forgotten this from our meeting discussions. Thanks for the reminder. I suggest to add a static method to the `StatsbeatModule` class. WDYT?

Comment (jeanbisutti) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorTraceExporter.java]: [I have added a method](https://github.com/Azure/azure-sdk-for-java/blob/e04961ebbc0f381c46d9597f8d88762efb054d1a/sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/FeatureStatsbeat.java#L54) in  `FeatureStatsbeat` to be able to add an instrumentation from a span.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorExporterBuilder.java [hunk 6]
```diff
@@ -99,6 +99,10 @@ public final class AzureMonitorExporterBuilder {
     // this is only populated after the builder is frozen
     private TelemetryItemExporter builtTelemetryItemExporter;
 
+    // this is only populated after the builder is frozen
+    private StatsbeatModule statsbeatModule;
+
+
```
Comment (trask) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/AzureMonitorExporterBuilder.java:105]:
```suggestion
```

**Review by trask — APPROVED** [review]
